README.md
# NAME

MooseX::Method::Signatures - Method declarations with type constraints and no source filter

# VERSION

version 0.47

# SYNOPSIS

    package Foo;

    use Moose;
    use MooseX::Method::Signatures;

    method morning (Str $name) {
        $self->say("Good morning ${name}!");
    }

    method hello (Str :$who, Int :$age where { $_ > 0 }) {
        $self->say("Hello ${who}, I am ${age} years old!");
    }

    method greet (Str $name, Bool :$excited = 0) {
        if ($excited) {
            $self->say("GREETINGS ${name}!");
        }
        else {
            $self->say("Hi ${name}!");
        }
    }

    $foo->morning('Resi');                          # This works.

    $foo->hello(who => 'world', age => 42);         # This too.

    $foo->greet('Resi', excited => 1);              # And this as well.

    $foo->hello(who => 'world', age => 'fortytwo'); # This doesn't.

    $foo->hello(who => 'world', age => -23);        # This neither.

    $foo->morning;                                  # Won't work.

    $foo->greet;                                    # Will fail.

# DESCRIPTION

Provides a proper method keyword, like "sub" but specifically for making methods
and validating their arguments against Moose type constraints.

# SIGNATURE SYNTAX

The signature syntax is heavily based on Perl 6. However not the full Perl 6
signature syntax is supported yet and some of it never will be.

## Type Constraints

    method foo (             $affe) # no type checking
    method bar (Animal       $affe) # $affe->isa('Animal')
    method baz (Animal|Human $affe) # $affe->isa('Animal') || $affe->isa('Human')

## Positional vs. Named

    method foo ( $a,  $b,  $c) # positional
    method bar (:$a, :$b, :$c) # named
    method baz ( $a,  $b, :$c) # combined

## Required vs. Optional

    method foo ($a , $b!, :$c!, :$d!) # required
    method bar ($a?, $b?, :$c , :$d?) # optional

## Defaults

    method foo ($a = 42) # defaults to 42

## Constraints

    method foo ($foo where { $_ % 2 == 0 }) # only even

## Invocant

    method foo (        $moo) # invocant is called $self and is required
    method bar ($self:  $moo) # same, but explicit
    method baz ($class: $moo) # invocant is called $class

## Labels

    method foo (:     $affe ) # called as $obj->foo(affe => $value)
    method bar (:apan($affe)) # called as $obj->foo(apan => $value)

## Traits

    method foo (Affe $bar does trait)
    method foo (Affe $bar is trait)

The only currently supported trait is `coerce`, which will attempt to coerce
the value provided if it doesn't satisfy the requirements of the type
constraint.

## Placeholders

    method foo ($bar, $, $baz)

Sometimes you don't care about some parameters you're being called with. Just put
the bare sigil instead of a full variable name into the signature to avoid an
extra lexical variable to be created.

## Complex Example

    method foo ( SomeClass $thing where { $_->can('stuff') }:
                 Str  $bar  = "apan",
                 Int :$baz! = 42 where { $_ % 2 == 0 } where { $_ > 10 } )

    # the invocant is called $thing, must be an instance of SomeClass and
           has to implement a 'stuff' method
    # $bar is positional, required, must be a string and defaults to "apan"
    # $baz is named, required, must be an integer, defaults to 42 and needs
    #      to be even and greater than 10

# BUGS, CAVEATS AND NOTES

This module is as stable now, but this is not to say that it is entirely bug
free. If you notice any odd behaviour (messages not being as good as they could
for example) then please raise a bug.

## Fancy signatures

[Parse::Method::Signatures](https://metacpan.org/pod/Parse::Method::Signatures) is used to parse the signatures. However, some
signatures that can be parsed by it aren't supported by this module (yet).

## No source filter

While this module does rely on the hairy black magic of [Devel::Declare](https://metacpan.org/pod/Devel::Declare) it
does not depend on a source filter. As such, it doesn't try to parse and
rewrite your source code and there should be no weird side effects.

Devel::Declare only effects compilation. After that, it's a normal subroutine.
As such, for all that hairy magic, this module is surprisingly stable.

## What about regular subroutines?

[Devel::Declare](https://metacpan.org/pod/Devel::Declare) cannot yet change the way `sub` behaves. However, the
[signatures](https://metacpan.org/pod/signatures) module can. Right now it only provides very basic
signatures, but it's extendable enough that plugging MooseX::Method::Signatures
signatures into that should be quite possible.

## What about the return value?

Type constraints for return values can be declared using

    method foo (Int $x, Str $y) returns (Bool) { ... }

however, this feature only works with scalar return values and is still
considered to be experimental.

## Interaction with [Moose::Role](https://metacpan.org/pod/Moose::Role)

### Methods not seen by a role's `requires`

Because the processing of the [MooseX::Method::Signatures](https://metacpan.org/pod/MooseX::Method::Signatures)
`method` and the [Moose](https://metacpan.org/pod/Moose) `with` keywords are both
done at runtime, it can happen that a role will require
a method before it is declared (which will cause
Moose to complain very loudly and abort the program).

For example, the following will not work:

    # in file Canine.pm

    package Canine;

    use Moose;
    use MooseX::Method::Signatures;

    with 'Watchdog';

    method bark { print "Woof!\n"; }

    1;



    # in file Watchdog.pm

    package Watchdog;

    use Moose::Role;

    requires 'bark';  # will assert! evaluated before 'method' is processed

    sub warn_intruder {
        my $self = shift;
        my $intruder = shift;

        $self->bark until $intruder->gone;
    }

    1;

A workaround for this problem is to use `with` only
after the methods have been defined.  To take our previous
example, __Canine__ could be reworked thus:

    package Canine;

    use Moose;
    use MooseX::Method::Signatures;

    method bark { print "Woof!\n"; }

    with 'Watchdog';

    1;

A better solution is to use [MooseX::Declare](https://metacpan.org/pod/MooseX::Declare) instead of plain
[MooseX::Method::Signatures](https://metacpan.org/pod/MooseX::Method::Signatures). It defers application of roles until the end
of the class definition. With it, our example would becomes:

    # in file Canine.pm

    use MooseX::Declare;

    class Canine with Watchdog {
        method bark { print "Woof!\n"; }
    }

    1;

    # in file Watchdog.pm

    use MooseX::Declare;

    role Watchdog {
        requires 'bark';

        method warn_intruder ( $intruder ) {
            $self->bark until $intruder->gone;
        }
    }

    1;

### _Subroutine redefined_ warnings

When composing a [Moose::Role](https://metacpan.org/pod/Moose::Role) into a class that uses
[MooseX::Method::Signatures](https://metacpan.org/pod/MooseX::Method::Signatures), you may get a "Subroutine redefined"
warning. This happens when both the role and the class define a
method/subroutine of the same name. (The way roles work, the one
defined in the class takes precedence.) To eliminate this warning,
make sure that your `with` declaration happens after any
method/subroutine declarations that may have the same name as a
method/subroutine within a role.

# CAVEATS

You are encouraged to read the ["WARNING" in MooseX::Declare](https://metacpan.org/pod/MooseX::Declare#WARNING).

# SEE ALSO

[MooseX::Declare](https://metacpan.org/pod/MooseX::Declare)

[Method::Signatures::Simple](https://metacpan.org/pod/Method::Signatures::Simple)

[Method::Signatures](https://metacpan.org/pod/Method::Signatures)

[Devel::Declare](https://metacpan.org/pod/Devel::Declare)

[Parse::Method::Signatures](https://metacpan.org/pod/Parse::Method::Signatures)

[Moose](https://metacpan.org/pod/Moose)

[signatures](https://metacpan.org/pod/signatures)

# AUTHOR

Florian Ragwitz <rafl@debian.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Florian Ragwitz.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# CONTRIBUTORS

- Ash Berlin <ash@cpan.org>
- Cory Watson <gphat@cpan.org>
- Dagfinn Ilmari Mannsåker <ilmari@ilmari.org>
- Daniel Ruoso <daniel@ruoso.com>
- Dave Rolsky <autarch@urth.org>
- Hakim Cassimally <osfameron@cpan.org>
- Jesse Luehrs <doy@tozt.net>
- Jonathan Rockway <jon@jrock.us>
- Jonathan Scott Duff <duff@pobox.com>
- Justin Hunter <justin.d.hunter@gmail.com>
- Karen Etheridge <ether@cpan.org>
- Kent Fredric <kentfredric@gmail.com>
- Lukas Mai <l.mai@web.de>
- Maik Hentsche <maik.hentsche@amd.com>
- Matt Kraai <kraai@ftbfs.org>
- Nicholas Perez <nperez@cpan.org>
- Rhesa Rozendaal <rhesa@cpan.org>
- Ricardo SIGNES <rjbs@cpan.org>
- Sebastian Willert <willert@cpan.org>
- Steffen Schwigon <ss5@renormalist.net>
- Yanick Champoux <yanick@babyl.dyndns.org>


Changes
Revision history for MooseX-Method-Signatures

{{$NEXT}}

0.49      2015-10-07 01:07:44Z
  - added x_deprecated to metadata

0.48      2014-09-20 17:46:45Z
  - fix tests that fail on altered warning messages in perl 5.21.4 (see perl
    RT#121638)

0.47      2013-12-14 19:06:59Z
  - update configure_requires checking in Makefile.PL, add CONTRIBUTING file

0.46      2013-10-02 00:40:37Z
  - re-release with fixed compile test

0.45      2013-09-12 03:06:25Z
  - removed use of deprecated Class::MOP::load_class
  - documentation reference added, pointing to the WARNING section in
    MooseX::Declare, regarding future utility of this module
  - repository migrated to the github moose organization

0.44      2012-11-24 21:17:02Z
  - Fix broken test in 5.17.6, where hash ordering is now randomized
    (RT#81388, Karen Etheridge)

0.43      2012-05-21 00:34:04Z
  - Bump dependency on MooseX::Types to 0.35 to resolve issues with recent
    MooseX::Types releases.

0.42      2012-04-23 18:08:19Z
  - Mark failing test due to Eval::Closure/Devel::Declare issues as TODO until
    things are fixed, to at least allow recent versions to install without
    errors.

0.41      2012-03-11 17:05:27Z
  - Add explicit dependency on Eval::Closure, in an effort to get more
    diagnostics from failing cpantesters reports.  No functionality changes.

0.40      2012-03-10 17:03:51Z
  - Loosen a regexp in a test, to handle the new format emitted by Carp 1.25.

0.39      2012-03-10 00:54:18Z
  - The test suite now uses Test::Fatal instead of Test::Exception (Karen
    Etheridge).
  - Bump dependency on MooseX::Types::Structured to 0.24, which contains
    more detailed error messages as well as some performance enhancements.

0.38      2012-03-09 23:43:07Z
  - Require B::Hooks::EndOfScope 0.10 and Module::Runtime 0.012, fixing various
    perl 5.8 breakages.

0.37      2011-09-12 09:35:29Z
  - Don't look in the stash for traits or types that contain :: (Jesse Luehrs).

0.36      2010-08-31 21:36:19Z
  - Avoid deprecation warnings from new Moose version by depending on
    MooseX::LazyRequire 0.06 (Jeremy Carman).

0.35      2010-07-19 03:50:08Z
  - Updated to avoid warnings from Moose 1.09 (Dave Rolsky).
  - Switch from Module::Install to Dist::Zilla.

0.34      2010-06-08 17:00:00Z
  - Provides a version of reify that replaces the original version and
    now finally CatalystX::Declare and MooseX::MultiMethods work.

0.33      2010-06-01 16:47:34Z
  - CatalystX::Declare seems to create a method object without a code
    reference at all.

0.32      2010-06-01 16:47:34Z
  - Detect usage of old MXMS::Meta::Method API and retro-compatibilize.

0.31      2010-06-01 16:47:34Z
  - Add support for "does" as a trait declarator
  - Compatibilize MXMS::Meta::Method signature with Class::MOP::Method
    so other traits can apply customizations to the actual body of the
    method.

0.30      2010-02-04 01:41:36Z
  - Make Any instead of Defined the default type constraint.
  - Add tests for coercing optional named parameters (Cory Watson).
    + Make those tests pass by depending on
      MooseX::Types::Structured 0.20.
  - Make sure Text::Balanced doesn't screw up our data.

0.29      2009-10-15 21:27:08Z
  - Fix regression introduced in 0.28.
    Validation for methods without named arguments wasn't strict enough
    anymore. Passing too many arguments hasn't been rejected. Thanks
    Justin Hunter for finding this.

0.28      2009-10-08 23:28:02Z
  - Fix default values for named arguments.

0.27      2009-09-22 12:30:23Z
  - Depend on MooseX::LazyRequire 0.04 to avoid fail on version 0.03, which has
    been broken by recent Moose versions (Closes RT#49903) (Maik Hentsche).

0.26      2009-09-08 18:14:01Z
  - Rebuilt dist to include missing files

0.25      2009-09-07 21:03:03Z
  - Remove scary ALPHA disclaimer from POD
  - Depend on Test::More 0.88 for done_testing

0.24      2009-08-31 17:04:14Z
  - Rename the meta method's clone method to reify. clone already had a special
    meaning for methods in role applications to classes.

0.23      2009-08-27 20:22:40Z
  - Bump Moose dependency to 0.89. This avoids test failures on some older
    Moose versions.

0.22      2009-08-27 13:29:25Z
  - Provide sane closure behaviour for anonymous methods.
  - Add tests for the above (Sebastian Willert).
  - Depend on MooseX::Types 0.19, as 0.18 made us blow up spectacularly.

0.21      2009-08-18 06:45:48Z
  - Add tests for using aliased.pm for type aliases (Steffen Schwigon).
  - Depend on the latest PMS to get type alias resolving.

0.20      2009-08-17 10:11:32Z
  - Make the meta method's parsed_signature public.
  - Use MooseX::LazyRequire's new interface and depend on 0.03.

0.19      2009-08-14 09:23:42Z
  - Move method redefinition warnings to the method application hook so it's
    easy to override.

0.18      2009-08-14 00:24:41Z
  - Depend on Devel::Declare 0.005011. This avoids "->method" and "method =>"
    being misinterpreted sometimes.

0.17      2009-08-13 06:37:31Z
  - Implement method traits (Cory Watson).
  - Properly test list placeholders.
  - Document the right syntax for return value type constraints.
  - Use namespace::autoclean instead of namespace::clean.
  - Let the signature default to (@) if none is given.
  - Don't require the package attribute when a pre-built context is passed
    in.
  - Depend on a recent Parse::Method::Signatures to make sure we will work
    with new PPIs and have the right MooseX::Traits version.
  - Fix doc typo (arcanez).
  - Split out context from self so that we may pass in an arbitrary context
    (nperez).
  - Enable passing a context and also a coderef to use to apply the method
    once constructed (for deferred method modifier application in MXD)
    (nperez).
  - Enable MXMS to work with multiple different declarators and prototype
    injections (nperez).
  - Change from using Devel::Declare::MethodInstaller::Simple to

lib/MooseX/Method/Signatures.pm
use strict;
use warnings;

package MooseX::Method::Signatures;
# ABSTRACT: (DEPRECATED) Method declarations with type constraints and no source filter
# KEYWORDS: moose extension method declaration signature prototype syntax sugar deprecated

our $VERSION = '0.50';

use Moose 0.89;
use Devel::Declare 0.005011 ();
use B::Hooks::EndOfScope 0.10;
use Moose::Meta::Class;
use MooseX::LazyRequire 0.06;
use MooseX::Types::Moose 0.19 qw/Str Bool CodeRef/;
use Text::Balanced qw/extract_quotelike/;
use MooseX::Method::Signatures::Meta::Method;
use MooseX::Method::Signatures::Types qw/PrototypeInjections/;
use Sub::Name;
use Moose::Util 'find_meta';
use Module::Runtime 'use_module';
use Carp;

use aliased 'Devel::Declare::Context::Simple', 'ContextSimple';

use namespace::autoclean;

has package => (
    is            => 'ro',
    isa           => Str,
    lazy_required => 1,
);

has context => (
    is      => 'ro',
    isa     => ContextSimple,
    lazy    => 1,
    builder => '_build_context',
);

has initialized_context => (
    is      => 'ro',
    isa     => Bool,
    default => 0,
);

has custom_method_application => (
    is        => 'ro',
    isa       => CodeRef,
    predicate => 'has_custom_method_application',
);

has prototype_injections => (
    is        => 'ro',
    isa       => PrototypeInjections,
    predicate => 'has_prototype_injections',
);

sub _build_context {
    my ($self) = @_;
    return ContextSimple->new(into => $self->package);
}

sub import {
    my ($class, %args) = @_;
    my $caller = caller();
    $class->setup_for($caller, \%args);
}

sub setup_for {
    my ($class, $pkg, $args) = @_;

    # process arguments to import
    while (my ($declarator, $injections) = each %{ $args }) {
        my $obj = $class->new(
            package              => $pkg,
            prototype_injections => {
                declarator => $declarator,
                injections => $injections,
            },
        );

        Devel::Declare->setup_for($pkg, {
            $declarator => { const => sub { $obj->parser(@_) } },
        });

        {
            no strict 'refs';
            *{ "${pkg}::$declarator" } = sub {};
        }
    }

    my $self = $class->new(package => $pkg);

    Devel::Declare->setup_for($pkg, {
        method => { const => sub { $self->parser(@_) } },
    });

    {
        no strict 'refs';
        *{ "${pkg}::method" } = sub {};
    }

    return;
}

sub strip_name {
    my ($self) = @_;
    my $ctx = $self->context;
    my $ret = $ctx->strip_name;
    return $ret if defined $ret;

    my $line = $ctx->get_linestr;
    my $offset = $ctx->offset;
    local $@;
    my $copy = substr($line, $offset);
    my ($str) = extract_quotelike($copy);
    return unless defined $str;

    return if ($@ && $@ =~ /^No quotelike operator found/);
    die $@ if $@;

    substr($line, $offset, length $str) = '';
    $ctx->set_linestr($line);

    return \$str;
}

sub strip_traits {
    my ($self) = @_;

    my $ctx = $self->context;
    my $linestr = $ctx->get_linestr;

    unless (substr($linestr, $ctx->offset, 2) eq 'is' ||
            substr($linestr, $ctx->offset, 4) eq 'does') {
        # No 'is' means no traits
        return;
    }

    my @traits;

    while (1) {
        if (substr($linestr, $ctx->offset, 2) eq 'is') {
            # Eat the 'is' so we can call strip_names_and_args
            substr($linestr, $ctx->offset, 2) = '';
        } elsif (substr($linestr, $ctx->offset, 4) eq 'does') {
            # Eat the 'does' so we can call strip_names_and_args
            substr($linestr, $ctx->offset, 4) = '';
        } else {
            last;
        }

        $ctx->set_linestr($linestr);
        push @traits, @{ $ctx->strip_names_and_args };
        # Get the current linestr so that the loop can look for more 'is'
        $ctx->skipspace;
        $linestr = $ctx->get_linestr;
    }

    confess "expected traits after 'is' or 'does', found nothing"
        unless scalar(@traits);

    # Let's check to make sure these traits aren't aliased locally
    for my $t (@traits) {
        next if $t->[0] =~ /::/;
        my $class = $ctx->get_curstash_name;
        my $meta = find_meta($class) || Moose::Meta::Class->initialize($class);
        my $func = $meta->get_package_symbol('&' . $t->[0]);
        next unless $func;

        my $proto = prototype $func;
        next if !defined $proto || length $proto;

        $t->[0] = $func->();
    }

    return \@traits;
}

sub strip_return_type_constraint {
    my ($self) = @_;
    my $ctx = $self->context;
    my $returns = $ctx->strip_name;
    return unless defined $returns;
    confess "expected 'returns', found '${returns}'"
        unless $returns eq 'returns';
    return $ctx->strip_proto;
}

sub parser {
    my $self = shift;
    my $err;

    # Keep any previous compile errors from getting stepped on. But report
    # errors from inside MXMS nicely.
    {
        local $@;
        eval { $self->_parser(@_) };
        $err = $@;
    }

    die $err if $err;
}

my $anon_counter = 1;
sub _parser {
    my $self = shift;
    my $ctx = $self->context;
    $ctx->init(@_) unless $self->initialized_context;

    $ctx->skip_declarator;
    my $name   = $self->strip_name;
    my $proto  = $ctx->strip_proto;
    my $attrs  = $ctx->strip_attrs || '';
    my $traits = $self->strip_traits;
    my $ret_tc = $self->strip_return_type_constraint;

    my $compile_stash = $ctx->get_curstash_name;

    my %args = (
      # This might get reset later, but its where we search for exported
      # symbols at compile time
      package_name => $compile_stash,
    );
    $args{ signature        } = qq{($proto)} if defined $proto;
    $args{ traits           } = $traits      if $traits;
    $args{ return_signature } = $ret_tc      if defined $ret_tc;

    # Class::MOP::Method requires a name
    $args{ name             } = $name || '__ANON__'.($anon_counter++).'__';

    if ($self->has_prototype_injections) {
        confess('Configured declarator does not match context declarator')
            if $ctx->declarator ne $self->prototype_injections->{declarator};
        $args{prototype_injections} = $self->prototype_injections->{injections};
    }

    my $meth_class = 'MooseX::Method::Signatures::Meta::Method';
    if ($args{traits}) {
        my @traits = ();
        foreach my $t (@{$args{traits}}) {
            use_module($t->[0]);
            if ($t->[1]) {
                %args = (%args, eval $t->[1]);
            };
            push @traits, $t->[0];
        }
        my $meta = Moose::Meta::Class->create_anon_class(
            superclasses => [ $meth_class  ],
            roles        => [ @traits ],
            cache        => 1,
        );
        $meth_class = $meta->name;
        delete $args{traits};
    }

    my $proto_method = $meth_class->wrap(sub { }, %args);

    my $after_block = ')';

    if ($traits) {
        if (my @trait_args = grep { defined } map { $_->[1] } @{ $traits }) {
            $after_block = q{, } . join(q{,} => @trait_args) . $after_block;
        }
    }

    if (defined $name) {
        my $name_arg = q{, } . (ref $name ? ${$name} : qq{q[${name}]});
        $after_block = $name_arg . $after_block . q{;};
    }

    my $inject = $proto_method->injectable_code;
    $inject = $self->scope_injector_call($after_block) . $inject;

    $ctx->inject_if_block($inject, "(sub ${attrs} ");

    my $create_meta_method = sub {
        my ($code, $pkg, $meth_name, @args) = @_;
        subname $pkg . "::" .$meth_name, $code;

        # we want to reinitialize with all the args,
        # so we give the opportunity for traits to wrap the correct
        # closure.
        my %other_args = %{$proto_method};
        delete $other_args{body};
        delete $other_args{actual_body};

        my $ret = $meth_class->wrap(
            $code,
            %other_args, @args
        );
    };

    if (defined $name) {
        my $apply = $self->has_custom_method_application
            ? $self->custom_method_application
            : sub {
                my ($meta, $name, $method) = @_;

                if (warnings::enabled("redefine") && (my $meta_meth = $meta->get_method($name))) {
                    warnings::warn("redefine", "Method $name redefined on package ${ \$meta->name }")
                        if $meta_meth->isa('MooseX::Method::Signatures::Meta::Method');
                }

                $meta->add_method($name => $method);
            };

        $ctx->shadow(sub {
            my ($code, $name, @args) = @_;

            my $pkg = $compile_stash;
            ($pkg, $name) = $name =~ /^(.*)::([^:]+)$/
                if $name =~ /::/;

            my $meth = $create_meta_method->($code, $pkg, $name, @args);
            my $meta = Moose::Meta::Class->initialize($pkg);

            $meta->$apply($name, $meth);
            return;
        });
    }
    else {
        $ctx->shadow(sub {
            return $create_meta_method->(shift, $compile_stash, '__ANON__', @_);
        });
    }
}

sub scope_injector_call {
    my ($self, $code) = @_;
    $code =~ s/'/\\'/g; # we're generating code that's quoted with single quotes
    return qq[BEGIN { ${\ref $self}->inject_scope('${code}') }];
}

sub inject_scope {
    my ($class, $inject) = @_;
    on_scope_end {
        my $line = Devel::Declare::get_linestr();
        return unless defined $line;
        my $offset = Devel::Declare::get_linestr_offset();
        substr($line, $offset, 0) = $inject;
        Devel::Declare::set_linestr($line);
    };
}

__PACKAGE__->meta->make_immutable;

1;

=head1 SYNOPSIS

    package Foo;

    use Moose;
    use MooseX::Method::Signatures;

    method morning (Str $name) {
        $self->say("Good morning ${name}!");
    }

    method hello (Str :$who, Int :$age where { $_ > 0 }) {
        $self->say("Hello ${who}, I am ${age} years old!");
    }

    method greet (Str $name, Bool :$excited = 0) {
        if ($excited) {
            $self->say("GREETINGS ${name}!");
        }
        else {
            $self->say("Hi ${name}!");
        }
    }

    $foo->morning('Resi');                          # This works.

    $foo->hello(who => 'world', age => 42);         # This too.

    $foo->greet('Resi', excited => 1);              # And this as well.

    $foo->hello(who => 'world', age => 'fortytwo'); # This doesn't.

    $foo->hello(who => 'world', age => -23);        # This neither.

    $foo->morning;                                  # Won't work.

    $foo->greet;                                    # Will fail.

=head1 DEPRECATION NOTICE

=for stopwords mst

=for comment rafl agreed we should have a warning, and mst wrote this for MooseX::Declare, but it applies equally well here:

B<Warning:> MooseX::Method::Signatures and L<MooseX::Declare> are based on
L<Devel::Declare>, a giant bag of crack originally implemented by mst with the
goal of upsetting the perl core developers so much by its very existence that
they implemented proper keyword handling in the core.

As of perl5 version 14, this goal has been achieved, and modules such as
L<Devel::CallParser>, L<Function::Parameters>, and L<Keyword::Simple> provide
mechanisms to mangle perl syntax that don't require hallucinogenic drugs to
interpret the error messages they produce.

If you want to use declarative syntax in new code, please for the love
of kittens get yourself a recent perl and look at L<Moops> and
L<core signatures|perlsub/Signatures> instead.

=head1 DESCRIPTION

Provides a proper method keyword, like "sub" but specifically for making methods
and validating their arguments against Moose type constraints.

=head1 SIGNATURE SYNTAX

The signature syntax is heavily based on Perl 6. However not the full Perl 6
signature syntax is supported yet and some of it never will be.

=head2 Type Constraints

    method foo (             $affe) # no type checking
    method bar (Animal       $affe) # $affe->isa('Animal')
    method baz (Animal|Human $affe) # $affe->isa('Animal') || $affe->isa('Human')

=head2 Positional vs. Named

    method foo ( $a,  $b,  $c) # positional
    method bar (:$a, :$b, :$c) # named
    method baz ( $a,  $b, :$c) # combined

=head2 Required vs. Optional

    method foo ($a , $b!, :$c!, :$d!) # required
    method bar ($a?, $b?, :$c , :$d?) # optional

=head2 Defaults

    method foo ($a = 42) # defaults to 42

=head2 Constraints

    method foo ($foo where { $_ % 2 == 0 }) # only even

=for stopwords Invocant

=head2 Invocant

    method foo (        $moo) # invocant is called $self and is required
    method bar ($self:  $moo) # same, but explicit
    method baz ($class: $moo) # invocant is called $class

=head2 Labels

    method foo (:     $affe ) # called as $obj->foo(affe => $value)
    method bar (:apan($affe)) # called as $obj->foo(apan => $value)

=head2 Traits

    method foo (Affe $bar does trait)
    method foo (Affe $bar is trait)

The only currently supported trait is C<coerce>, which will attempt to coerce
the value provided if it doesn't satisfy the requirements of the type
constraint.

=head2 Placeholders

    method foo ($bar, $, $baz)

=for stopwords sigil

Sometimes you don't care about some parameters you're being called with. Just put
the bare sigil instead of a full variable name into the signature to avoid an
extra lexical variable to be created.

=head2 Complex Example

    method foo ( SomeClass $thing where { $_->can('stuff') }:
                 Str  $bar  = "apan",
                 Int :$baz! = 42 where { $_ % 2 == 0 } where { $_ > 10 } )

    # the invocant is called $thing, must be an instance of SomeClass and
           has to implement a 'stuff' method
    # $bar is positional, required, must be a string and defaults to "apan"
    # $baz is named, required, must be an integer, defaults to 42 and needs
    #      to be even and greater than 10

=head1 CAVEATS AND NOTES

This module is as stable now, but this is not to say that it is entirely bug
free. If you notice any odd behaviour (messages not being as good as they could
for example) then please raise a bug.

=head2 Fancy signatures

L<Parse::Method::Signatures> is used to parse the signatures. However, some
signatures that can be parsed by it aren't supported by this module (yet).

=head2 No source filter

While this module does rely on the hairy black magic of L<Devel::Declare> it
does not depend on a source filter. As such, it doesn't try to parse and
rewrite your source code and there should be no weird side effects.

Devel::Declare only effects compilation. After that, it's a normal subroutine.
As such, for all that hairy magic, this module is surprisingly stable.

=head2 What about regular subroutines?

L<Devel::Declare> cannot yet change the way C<sub> behaves. However, the
L<signatures|signatures> module can. Right now it only provides very basic
signatures, but it's extendable enough that plugging MooseX::Method::Signatures
signatures into that should be quite possible.

=head2 What about the return value?

Type constraints for return values can be declared using

  method foo (Int $x, Str $y) returns (Bool) { ... }

however, this feature only works with scalar return values and is still
considered to be experimental.

=head2 Interaction with L<Moose::Role>

=head3 Methods not seen by a role's C<requires>

Because the processing of the L<MooseX::Method::Signatures>
C<method> and the L<Moose> C<with> keywords are both
done at runtime, it can happen that a role will require
a method before it is declared (which will cause
Moose to complain very loudly and abort the program).

For example, the following will not work:

    # in file Canine.pm

    package Canine;

    use Moose;
    use MooseX::Method::Signatures;

    with 'Watchdog';

    method bark { print "Woof!\n"; }

    1;


    # in file Watchdog.pm

    package Watchdog;

    use Moose::Role;

    requires 'bark';  # will assert! evaluated before 'method' is processed

    sub warn_intruder {
        my $self = shift;
        my $intruder = shift;

        $self->bark until $intruder->gone;
    }

    1;


A workaround for this problem is to use C<with> only
after the methods have been defined.  To take our previous
example, B<Canine> could be reworked thus:

    package Canine;

    use Moose;
    use MooseX::Method::Signatures;

    method bark { print "Woof!\n"; }

    with 'Watchdog';

    1;


A better solution is to use L<MooseX::Declare> instead of plain
L<MooseX::Method::Signatures>. It defers application of roles until the end
of the class definition. With it, our example would becomes:


    # in file Canine.pm

    use MooseX::Declare;

    class Canine with Watchdog {
        method bark { print "Woof!\n"; }
    }

    1;

    # in file Watchdog.pm

    use MooseX::Declare;

    role Watchdog {
        requires 'bark';

        method warn_intruder ( $intruder ) {
            $self->bark until $intruder->gone;
        }
    }

    1;


=head3 I<Subroutine redefined> warnings

When composing a L<Moose::Role> into a class that uses
L<MooseX::Method::Signatures>, you may get a "Subroutine redefined"
warning. This happens when both the role and the class define a
method/subroutine of the same name. (The way roles work, the one
defined in the class takes precedence.) To eliminate this warning,
make sure that your C<with> declaration happens after any
method/subroutine declarations that may have the same name as a
method/subroutine within a role.

=head1 SEE ALSO

=for :list
* L<MooseX::Declare>
* L<Method::Signatures::Simple>
* L<Method::Signatures>
* L<Devel::Declare>
* L<Parse::Method::Signatures>
* L<Moose>
* L<signatures>

=cut
